### ISSUE TITLE
React hooks array destructuring

### ISSUE THREAD

**Issue by gnordhielm** [prettier/prettier#5851]
Following up on this discussion: https://github.com/prettier/prettier/issues/3950.

I'll reiterate what @sompylasar said - arrays and objects should behave the same in the same scenarios.

The contributor mentioned that this was a behavior that you guys changed because of user feedback. I think you should consider changing it back because of user feedback. Array destructuring is central to the React hooks api, and it's rather frustrating to have this:

`const [ isLoading, setIsLoading ] = useState(false)`

Changed to this:

`const [isLoading, setIsLoading] = useState(false)`

I imagine more users will have a similar complaint as hooks are more widely adopted. Perhaps this should be re-evaluated? If there is strong disagreement about this behavior perhaps it's a candidate for an option?

And, of course thank you for putting together such a thoughtfully designed library. It really does make my job so much easier. I'm also happy to submit a PR for the change if it's helpful.

**Comment by gnordhielm** [comment]
@ikatyang @lydell @j-f1

**Comment by lydell** [comment]
@gnordhielm No need to ping us. Let's wait for your proposal to make traction.

This is the first example in https://reactjs.org/docs/hooks-intro.html:

```js
import React, { useState } from 'react';

function Example() {
  // Declare a new state variable, which we'll call "count"
  const [count, setCount] = useState(0);

  return (
    <div>
      <p>You clicked {count} times</p>
      <button onClick={() => setCount(count + 1)}>
        Click me
      </button>
    </div>
  );
}
```

No spaces.

**Comment by vtambourine** [comment]
Personally, I would prefer better consistency in such scenarios. Seen two kinds of destructuring line by line feels weird:
```js
const { buttons } = wrapper.state();
const [firstButton, secondButton, thirdButton] = buttons;
```
Same spaces patters looks better and raises to question:
```js
const { buttons } = wrapper.state();
const [ firstButton, secondButton, thirdButton ] = buttons;
```

**Comment by gnordhielm** [comment]
I totally agree. It seems like deviating from consistency should come with some kind of explanation, and no one is offering one.

**Comment by j-f1** [comment]
Here’s an analysis of several popular styles, of which none recommend adding spaces there: https://github.com/prettier/prettier/issues/296#issuecomment-274840112

**Comment by vtambourine** [comment]
The analysis you linked looks into spaces used around brackets in array literals. The destructuring case is different. In destructuring, the `[ a, b, ...c ]` syntax doesn't represent the array, but _assignment pattern_. Therefore it's not strictly valid to apply same rules and logic that used to describe arrays style rules.

**Comment by bionicles** [comment]
for some reason, prettier breaks my useEffect array argument:

```
  useEffect(
    () => {
      updateSimulation();
    },
    [state]
  );
```
becomes
```
  useEffect(
    () => {
      updateSimulation();
    },
    []
  );
```
thus, prettier does not format my code; it changes the actual meaning, and severely

**Comment by hawkrives** [comment]
@bionicles Can you reproduce your issue on the prettier.io playground? If not, can you reproduce it with just a command-line invocation of `prettier`? 

What version of Prettier are you using? How are you invoking it?

**Comment by j-f1** [comment]
It looks like that might be part of [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks).

**Comment by duailibe** [comment]
@bionicles That has nothing do with Prettier, as you can see [checking in the playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBXAznAogMx3MGACgB0oACcogSnIF4A+c4Myy1ABwBMBDeAZQCWAW1QAbPoOg0A3K3IBfADTyA2uhh84AXTLUZIJSAgcYUqOmSgeAJxsQA7gAVbCSyh4A3CIK6GQAEY2PGAA1nAw-BwhglAA5sgwNqhwRgAWMMJiAOppgvDo0WBw-G75gp75AJ7I4OiWRrGYNjBOwXHCPMg4PGKYRgBW6AAeAELBYRH8PMJwADKxcN29-SBDw0LxYnAAiqgQ8Mt9qSDRNs21ATwBVWLQ-hw2sTDZvjBpyAAcAAxGjxCYbLBDi1R5wZqeJZGACO+3gbRM7hAPHQAFooHA4Fwsf4bHBYYI8W0eB0ukgescjJhhIJEskTuhYnFtnsDktySsTpoAq8uO9kAAmIxJHiCMRMgDCEGEnVqUGgULQmAAKtd3BTMAoFEA)

Besides, even though it's related to React hooks, it has nothing to do with the subject of this issue (array destructuring). In the future, open a new issue instead of commenting in an unrelated issue, that makes it harder for us to provide support for it.

Thanks
